syntax = "proto3";

package inventory;

service InventoryService {
    rpc SearchByID (SearchByIDRequest) returns (stream InventoryRecord);
    rpc SearchBySKU (SearchBySKURequest) returns (stream InventoryRecordList);
    rpc SearchBySKURange (SearchBySKURangeRequest) returns (stream InventoryRangeList);
    rpc GetDistribution (GetDistributionRequest) returns (stream InventoryRecordList);
    rpc UpdateCost (UpdateCostRequest) returns (stream UpdateResponse);
}

message SearchByIDRequest {
    string sku = 1;
}

message SearchBySKURequest {
    string key_name = 1;
    string key_value = 2;
}

message SearchBySKURangeRequest {
    string start_sku = 1;
    string end_sku = 2;
}

message GetDistributionRequest {
    double percentile = 1;
}

message UpdateCostRequest {
    string new_sku = 1;
    double new_cost = 2;
    bool success = 3;
}

message InventoryRecord {
    string sku = 1;
    string description = 2;
    string bin_number = 3;
    string location = 4;
    string unit = 5;
    int32 qty = 6;
    int32 reorder_qty = 7;
    double cost = 8;
}

message InventoryRecordList {
    string sku = 1;
    string description = 2;
    string bin_number = 3;
    string location = 4;
    string unit = 5;
    int32 qty = 6;
    int32 reorder_qty = 7;
    double cost = 8;
    string key_name = 9;
    string key_value = 10;

}

message InventoryRangeList {
    string start_sku = 1;
    string end_sku = 2;
    string sku = 3;
   }

message UpdateResponse {
    string new_sku = 1;
    double new_cost = 2;
    bool success = 3;
}
